typing/printtyp.ml

```diff
@@ -2436,6 +2436,13 @@ let explain_object (type variety) : variety Errortrace.obj -> _ = function
   | Errortrace.Self_cannot_be_closed ->
       Some (dprintf "@,Self type cannot be unified with a closed object type")
 
+let explain_incompatible_fields name (diff: Types.type_expr Errortrace.diff) =
+  dprintf "@,@[The method %a has type@ %a,@ \
+  but the expected method type was@ %a@]"
+    Style.inline_code name
+    (Style.as_inline_code type_expr_with_reserved_names) diff.got
+    (Style.as_inline_code type_expr_with_reserved_names) diff.expected
```

Comment:
The `type_expr_with_reserved_names` function should be called on type expressions whose explicit type variable names have been reserved.

This is not the case in the new branch, I propose to move the
```ocaml
  reserve_names diff.got;
  reserve_names diff.expected;
```
part in this function to avoid this.

Reply:
Makes sense, done in https://github.com/ocaml/ocaml/pull/12777/commits/392a30c3c235860d2c893759ead6ef7d37d867ad.

Just curious, and I assume this is out of the scope, but would it make sense to move the call to `reserve_names` inside `type_expr_with_reserved_names` itself?

Reply:
This would be suboptimal: `reserve_names` (and `prepare_types`) compute metadata to decide how to print loops in type expression and how to disambiguate type variable names. 

In both case, it is better to have the full picture available before starting to print.

Otherwise we might decide for instance to name `'b` an anonymous type variable in the first printed type expression whereas in the second the user was using `'b` explicitly and we end-up needed to rename the user-chosen variable names to account for our greedy choice of anonymous type variables.